### PR TITLE
Add GUC to re-enable YYYYMMDDHH24MISS timestamp support

### PR DESCRIPTION
**Pull request by dgkimura**
Commit 71ec37a9136 removed the implicit format in favor of better
aligning with Postgres. However, it broke existing customers upgrades
from 5X to 6X. This patch brings back the old behavior behind a GUC
'enable_implicit_timeformat_YYYYMMDDHH24MISS' that is disabled by
default.

Co-authored-by: Ashwin Agrawal <aashwin@vmware.com>

### CODE REVIEW

File: src/include/utils/unsync_guc_name.h [hunk 1]
```diff
@@ -534,3 +534,5 @@
 		"xmloption",
 		"zero_damaged_pages",
 		"max_slot_wal_keep_size",
+		"optimizer_use_external_constant_expression_evaluation_for_ints",
```
Comment (ashwinstar) [inline src/include/utils/unsync_guc_name.h:537]: Seems additional GUC got added here.

File: src/backend/utils/adt/datetime.c [hunk 2]
```diff
@@ -2900,9 +2901,9 @@ DecodeNumberField(int len, char *str, int fmask,
 		len = strlen(str);
 	}
 	/* No decimal point and no complete date yet? */
-	else if ((fmask & DTK_DATE_M) != DTK_DATE_M)
+	if ((fmask & DTK_DATE_M) != DTK_DATE_M)
```
Comment (ashwinstar) [inline src/backend/utils/adt/datetime.c:2904]: looking at the change now, feels we can keep the `else` here

File: src/include/utils/unsync_guc_name.h (hunk 1 above)
Comment (bhuvnesh2703) [reply src/include/utils/unsync_guc_name.h:537]: yep needs to be removed.

File: src/backend/utils/misc/guc_gp.c [hunk 3]
```diff
@@ -244,6 +244,7 @@ bool		gp_maintenance_mode;
 bool		gp_maintenance_conn;
 bool		allow_segment_DML;
 bool		gp_allow_rename_relation_without_lock = false;
+bool		enable_implicit_timeformat_YYYYMMDDHH24MISS;
```
Comment (bhuvnesh2703) [inline src/backend/utils/misc/guc_gp.c:247]: should it be enable_implicit_timeformat_conversion
the help text can have details of the format..

File: src/backend/utils/adt/datetime.c [hunk 4]
```diff
@@ -2902,7 +2903,7 @@ DecodeNumberField(int len, char *str, int fmask,
 	/* No decimal point and no complete date yet? */
 	else if ((fmask & DTK_DATE_M) != DTK_DATE_M)
 	{
-		if (len >= 6)
+		if (len >= 6 && len != 14)
```
Comment (bhuvnesh2703) [inline src/backend/utils/adt/datetime.c:2906]: to have the old behavior and would you want to check len != 14 only when the enable_implicit_timeformat_YYYYMMDDHH24MISS is set?

**Review by chrishajas — COMMENTED** [review]
LGTM after adding the GUC check to preserve exisitng behavior as @bhuvnesh2703 mentioned

File: src/backend/utils/adt/datetime.c (hunk 4 above)
Comment (ashwinstar) [reply src/backend/utils/adt/datetime.c:2906]: yes we are adding GUC check here based on commit, year can be any longer than 5+ digits :-)
```
commit 7778ddc7a2d5b006edbfa69cdb44b8d8c24ec1ff
Author: Bruce Momjian <bruce@momjian.us>
Date:   Wed Oct 16 13:22:55 2013 -0400

    Allow 5+ digit years for non-ISO timestamp/date strings, where appropriate
    
    Report from Haribabu Kommi
```

**Review by kainwen — COMMENTED** [review]
LGTM. Only concern from me:

1. we do not have case cover the transform on QEs (note constant exp might be evaluated in QD before dispatch). But if the GUC is sync-ed, seems reasonable not to add such.

2. The text format need to transform always have `len = 14`?

**Comment by ashwinstar** [comment]
> The text format need to transform always have len = 14?

We made it to have the behavior it was in GPDB5 and that's how the condition in gpdb5 is, so yes

**Pull request merged**